src/abstractdataframe/show.jl

```diff
@@ -597,50 +597,64 @@ function _show(io::IO,
                summary::Bool = true,
                eltypes::Bool = true,
-               rowid=nothing,
-               truncstring::Int)
+               truncstring::Int = 32,
+               kwargs...)
+
     _check_consistency(df)
 
-    # we will pass around this buffer to avoid its reallocation in ourstrwidth
-    buffer = IOBuffer(Vector{UInt8}(undef, 80), read=true, write=true)
+    names = permutedims(propertynames(df))
+    types = permutedims(compacttype.(eltype.(eachcol(df))))
 
-    nrows = size(df, 1)
-    if rowid !== nothing
-        if size(df, 2) == 0
-            rowid = nothing
-        elseif nrows != 1
-            throw(ArgumentError("rowid may be passed only with a single row data frame"))
-        end
-    end
-    dsize = displaysize(io)
-    availableheight = dsize[1] - 7
-    nrowssubset = fld(availableheight, 2)
-    bound = min(nrowssubset - 1, nrows)
-    if allrows || nrows <= availableheight
-        rowindices1 = 1:nrows
-        rowindices2 = 1:0
-    else
-        rowindices1 = 1:bound
-        rowindices2 = max(bound + 1, nrows - nrowssubset + 1):nrows
+    crop = :both
+
+    if allcols && allrows
+        crop = :none
+    elseif allcols
+        crop = :vertical
+    elseif allrows
+        crop = :horizontal
     end
-    maxwidths = getmaxwidths(df, io, rowindices1, rowindices2, rowlabel, rowid,
-                             eltypes, buffer, truncstring)
-    width = getprintedwidth(maxwidths)
-    showrows(io, df, rowindices1, rowindices2, maxwidths, splitcols, allcols,
-             rowlabel, summary, eltypes, rowid, buffer, truncstring)
-    return
+
+    # Check if the user wants to display a summary about the DataFrame that is
+    # being printed. This will be shown using the `title` option of
+    # `pretty_table`.
+    title = summary ? Base.summary(df) : ""
+
+    # Create the formatter considering the current maximum size of the strings.
+    _formatter = (v,i,j)->_df_formatter(v, i, j, truncstring)
+
+    # Print the table with the selected options.
+    pretty_table(io, df, vcat(names,types);
+                 alignment                   = :l,
+                 continuation_row_alignment  = :l,
+                 crop                        = crop,
+                 crop_num_lines_at_beginning = 2,
+                 formatters                  = (_formatter,),
+                 highlighters                = (_DF_HIGHLIGHTER,),
+                 maximum_columns_width       = truncstring,
+                 newline_at_end              = false,
+                 nosubheader                 = !eltypes,
+                 row_number_alignment        = :l,
+                 row_number_column_title     = string(rowlabel),
+                 show_row_number             = true,
+                 tf                          = dataframe,
+                 title                       = title,
+                 vlines                      = [1],
+                 kwargs...)
```

Comment:
are there other kwargs that can be reasonably passed?

Reply:
Yes, there are tons of configurations that the user can select if the way. Those are all possible keywords so far for the text printing:

```julia
             alignment::Union{Symbol,Vector{Symbol}} = :r,
             backend::Union{Nothing,Symbol} = nothing,
             cell_alignment::Union{Nothing,
                                   Dict{Tuple{Int,Int},Symbol},
                                   Function,
                                   Tuple} = nothing,
             cell_first_line_only::Bool = false,
             compact_printing::Bool = true,
             filters_row::Union{Nothing,Tuple} = nothing,
             filters_col::Union{Nothing,Tuple} = nothing,
             formatters::Union{Nothing,Function,Tuple} = nothing,
             header_alignment::Union{Symbol,Vector{Symbol}} = :s,
             header_cell_alignment::Union{Nothing,
                                          Dict{Tuple{Int,Int},Symbol},
                                          Function,
                                          Tuple} = nothing,
             row_names::Union{Nothing,AbstractVector} = nothing,
             row_name_alignment::Symbol = :r,
             row_name_column_title::AbstractString = "",
             row_number_column_title::AbstractString = "Row",
             show_row_number::Bool = false,
             title::AbstractString = "",
             title_alignment::Symbol = :l,
                  border_crayon::Crayon = Crayon(),
                  header_crayon::Union{Crayon,Vector{Crayon}} = Crayon(bold = true),
                  subheader_crayon::Union{Crayon,Vector{Crayon}} = Crayon(foreground = :dark_gray),
                  rownum_header_crayon::Crayon = Crayon(bold = true),
                  text_crayon::Crayon = Crayon(),
                  autowrap::Bool = false,
                  body_hlines::Vector{Int} = Int[],
                  body_hlines_format::Union{Nothing,NTuple{4,Char}} = nothing,
                  continuation_row_alignment::Symbol = :c,
                  crop::Symbol = :both,
                  crop_subheader::Bool = false,
                  crop_num_lines_at_beginning::Int = 0,
                  columns_width::Union{Int,AbstractVector{Int}} = 0,
                  equal_columns_width::Bool = false,
                  highlighters::Union{Highlighter,Tuple} = (),
                  hlines::Union{Nothing,Symbol,AbstractVector} = nothing,
                  linebreaks::Bool = false,
                  maximum_columns_width::Union{Int,AbstractVector{Int}} = 0,
                  minimum_columns_width::Union{Int,AbstractVector{Int}} = 0,
                  newline_at_end::Bool = true,
                  overwrite::Bool = false,
                  noheader::Bool = false,
                  nosubheader::Bool = false,
                  row_name_crayon::Crayon = Crayon(bold = true),
                  row_name_header_crayon::Crayon = Crayon(bold = true),
                  row_number_alignment::Symbol = :r,
                  screen_size::Union{Nothing,Tuple{Int,Int}} = nothing,
                  tf::TextFormat = unicode,
                  title_autowrap::Bool = false,
                  title_crayon::Crayon = Crayon(bold = true),
                  title_same_width_as_table::Bool = false,
                  vlines::Union{Nothing,Symbol,AbstractVector} = nothing)
```

Reply:
also have you checked what happens if someone passes a kwarg that clashes with the kwargs you set explicitly?

Reply:
In this case, the user option has precedence. Thus, they can really change anything they want.

Reply:
```julia
julia> df = DataFrame(a = 1 , b = 2)
1×2 DataFrame
 Row │ a      b
     │ Int64  Int64
─────┼──────────────
 1   │ 1      2

julia> show(df, vlines = :all)
1×2 DataFrame
│ Row │ a     │ b     │
│     │ Int64 │ Int64 │
├─────┼───────┼───────┤
│ 1   │ 1     │ 2     │
julia>
```